docs/painless/painless-casting.asciidoc

```diff
@@ -1,172 +1,472 @@
-occur automatically, such as during an assignment operation. Explicit casts are
-casts where you use the casting operator to explicitly convert one type to
-another. This is necessary during operations where the cast cannot be inferred.
+A cast is the conversion of the value of one type to the equivalent value of an
+inferred or specified type. Implicit casts are an inferred operation that
+automatically convert values between types within a statement or
+<<painless-operators, expression>> when necessary. Explicit casts are a
+specified operation that forcefully convert values between types and are
+required when an appropriate type for a statement or
+<<painless-operators, expression>> cannot be inferred.  If there exists no
+equivalent value for the type converted to during a cast operation an error
+will occur.  If an implicit cast is given but an explicit cast is required an
+error will occur.
 
-To cast to a new type, precede the expression by the new type enclosed in
-parentheses, for example
-`(int)x`.
-
-The following sections specify the implicit casts that can be performed and the
-explicit casts that are allowed. The only other permitted cast is casting
-a single character `String` to a `char`.
-
-*Grammar:*
+*Grammar*
 [source,ANTLR4]
 ----
 cast: '(' TYPE ')' expression
 ----
 
-[[numeric-casting]]
-==== Numeric Casting
+*Examples*
+
+* Valid general casts.
++
+[source,Painless]
+----
+<1> int i = (int)5L;
+<2> Map m = new HashMap();
+<3> HashMap hm = (HashMap)m;
+----
++
+<1> declare `int i`;
+    explicit cast value of `long 5` to value of `int 5` -> `int 5`;
+    assign value of `int 5` to `i`
+<2> declare `Map m`;
+    allocate `HashMap` instance -> `HashMap reference`;
+    implicit cast value of `HashMap reference` to value of `Map reference`
+            -> `Map reference`;
+    assign value of `Map reference` to `m`
+<3> declare `HashMap hm`;
+    access value of `m` -> `Map reference`;
+    explicit cast value of `Map reference` to value of `HashMap reference`
+            -> `HashMap reference`;
+    assign value of `HashMap reference` to `hm`
+
+[[numeric-type-casting]]
+==== Numeric Type Casting
 
-The following table shows the allowed implicit and explicit casts between
-numeric types. Read the table by row. To find out if you need to explicitly
-cast from type A to type B, find the row for type A and scan across to the
-column for type B.
+A <<primitive-types, numeric type>> cast is the conversion of the value of one
+<<primitive-types, numeric type>> to the equivalent value of an inferred or
+specified <<primitive-types, numeric type>>. Casts between values of
+<<primitive-types, numeric types>> will result in data loss when the value of
+the type converted from is larger than the equivalent value of the type
+converted to can accommodate. Casts between values of integral types and values
+of floating point types will result in precision loss when the value of the
```

Comment:
Maybe, "larger than the target type can accommodate"?

I'd drop the "wills" -- they do result in data/precision loss in those cases.

Maybe just "Casts between integer and floating point values can result in precision loss."?

Reply:
Fixed.